### ISSUE TITLE
Running into infinite loop with recursive schema

### ISSUE THREAD

**Issue by sw23** [coveooss/json-schema-for-humans#67]
I'm running into infinite recursion with a schema that looks like this:

```
{
    "$id": "https://example.com/person.schema.json",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Person",
    "type": "object",
    "properties": {
        "relationships": {
            "$ref": "#/definitions/person/properties/relationships"
        }
    },
    "definitions": {
        "person": {
            "type": "object",
            "description": "A human being",
            "properties": {
                "relationships": {
                    "type": "object",
                    "description": "Relationships between this person and others",
                    "properties": {
                        "mother": {
                            "$ref": "#/definitions/person"
                        }
                    }
                }
            }
        }
    }
}
```
What I'd expect is for a document that looks something like this:
- relationships
  - mother
    - relationships
      -> mother (link to existing mother section, because reference has been encountered before)

I've poked around the code a bit but haven't been able to get to the bottom of the issue yet

**Comment by dblanchette** [comment]
Hi!

Sorry for this, this issue keeps popping up in one form or another every time I think I fixed it for good (and the code gets more and more complex).

I'll take a look soon!

**Comment by dblanchette** [comment]
Fix released in version v0.27.2

How it works, if you're curious: The first step when rendering a schema is to build the intermediate representation. This is a Python object that represents the input JSON Schema in an easier to exploit format. The Jinja templates then uses that object. This avoids having to code complex logic into the templates, like resolving references and such.

When building the intermediate representation, each time a new `$ref` is encountered, it is added to a mapping. The mapping is followed in reverse order and if we are able to reach the current_node through itself, the current node is marked as not displayed.

In the template, we display a link instead of rendering the node when the `is_displayed` attribute from the node object is True. 

Here, I changed the object to be aware of the properties, additionalProperties and patternProperties directly, instead of having another node object whose sole purpose is to list properties.

I must admit I'm not 100% sure why this fixed the issue and I need to put some more time in to make sure it covers all use cases (no test failures is already a good look), but I like this new structure since it is lighter and simpler.
